### ISSUE TITLE
When using Distributed RPC Framework, there is a leaking memory

### ISSUE THREAD

**Issue by hhwode** [pytorch/pytorch#34359]
❓ Questions and Help
Hi.
When using Distributed RPC Framework, there is a leaking memory, I don't know if there are any pieces that need to be configured
### Description
When I ran the RNN example provided on  [RNN](https://github.com/pytorch/examples/tree/master/distributed/rpc/rnn), when the number of iterations reached 10,000, the RNN example slowly consumed all the memory. Not clear why. 
You can find the corresponding code from this link:

I trace the code and locate errors in this line:
```
emb = _remote_method(EmbeddingTable.forward, self.emb_table_rref, input)
***
decoded = _remote_method(Decoder.forward, self.decoder_rref, output)
```
It seems to be creating objects at remote nodes after one iteration.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jjlilley

**Comment by mrshenli** [comment]
@hhwode thanks for reporting. We will investigate. It could be Python GC didn't launch in time, or dist autograd context wasn't deleted fast enough on remote node, or a bug. 

For RRef, we can try explicitly call `del rref`. 
For dist autograd, do we have API to block waiting for dist autograd context deletion? @pritamdamania87

**Comment by mrshenli** [comment]
Hey @hhwode which version of PyTorch are you using. If it is v1.4, could you please try master or the nightly? We fixed some leaks after v1.4, e.g.,  #31030.

**Comment by rohan-varma** [comment]
> For dist autograd, do we have API to block waiting for dist autograd context deletion? @pritamdamania87

We don't and performance wise, I'm not sure if we would want it, as there is a new dist autograd that needs to be created + cleared for every iteration. For a blocking wait for deletion, we'd have to pause during every training iteration which would significantly hurt the performance

**Comment by jjlilley** [comment]
fwiw: there is a known memory leak in nightly/master, for which a fix is pending review. (#36502 )

**Comment by hhwode** [comment]
> Hey @hhwode which version of PyTorch are you using. If it is v1.4, could you please try master or the nightly? We fixed some leaks after v1.4, e.g., #31030.

I'm using PyTorch 1.4.0 and I'll try other versions.

**Comment by pritamdamania87** [comment]
Closing this out since we've had several fixes go in for this. @hhwode feel free to re-open the issue if you're still seeing some issues.
